server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -1441,6 +1441,10 @@ public final boolean shouldProcessCheckpoint(ReplicationCheckpoint requestCheckp
             logger.warn("Ignoring new replication checkpoint - shard is in primaryMode and cannot receive any checkpoints.");
             return false;
         }
+        if (this.routingEntry().primary()) {
+            logger.warn("Ignoring new replication checkpoint - primary shard cannot receive any checkpoints.");
+            return false;
+        }
```

Comment:
How does this check differs from `getReplicationTracker().isPrimaryMode()` check ? Can these two be clubbed together ?

Reply:
`routingEntry().primary()` is a cluster level state check, whereas `getReplicationTracker().isPrimaryMode()` is a node level check. With some integration tests we are hitting a scenario during relocation where a shard which is not in PrimaryMode but at a cluster level (Shard Routing) that shard is still acting as primary. To be precise this happens during relocation of old primary to new primary shard, after handoff local shards (node level) are updated but cluster state is not updated yet. 

So we add this PR's check to make sure old primary (which is still primary until cluster state is updated) doesn't process any checkpoint.

Reply:
When you mean clubbing both of them together you mean checking both conditions in same if block? something like this: `if (this.getReplicationTracker().isPrimaryMode() || this.routingEntry().primary()) {}`. If Yes, I think we write them together we just need to change log statement to reflect both

Reply:
The issue here is that the shardRouting entry drives if the engine is created as writable or of type NRTReplicationEngine, updates to this field are also used to determine when to change a shard's engine.  During relocation or failover the old primary would not be directly converted to a replica, it would be shut down & recovered entirely.   In addition to checking the routing here, I think we may want an assertion on the presence of NRTReplication engine in calling `getReplicationEngine()` ensuring the replica can support processing a checkpoint.

Reply:
Thanks @Rishikesh1159 @mch2 for the clarification. @Rishikesh1159 feel free to add assertion on engine type.